nltk/metrics/distance.py

```diff
@@ -366,6 +366,160 @@ def jaro_winkler_similarity(s1, s2, p=0.1, max_l=4):
+    :type s2: str
+    :rtype int
+
+    """
+
+    n1, n2 = len(s1), len(s2)
+    if min([len(s1), len(s2)]) == 0:
+        return max([len(s1), len(s2)])
+    else:
+        if s1[len(s1)-1] == s2[len(s2)-1]:
+            cost = 0
+        else:
+            cost = 1
+
+        dist = min(
+            [damerau_levenshtein_distance(s1=s1[:len(s1) - 1],
+                                          s2=s2) + 1,
+             damerau_levenshtein_distance(s1=s1,
+                                          s2=s2[:len(s2) - 1]) + 1,
+             damerau_levenshtein_distance(s1=s1[:len(s1) - 1],
+                                          s2=s2[:len(s2) - 1]) + cost])
+
+        if min([n1, n2]) >= 2 and s1[n1] == s2[n2-1] and s1[n1-1] == s2[n2-1]:
+            return min([dist, damerau_levenshtein_distance(s1=s1[:len(s1)-2],
+                                                           s2=s2[:len(s2)-2]) + cost])
+        else:
+            return dist
+
+
+def hamming_distance(s1, s2):
+    """
+    The Hamming distance measures the distance between two equal length strings
+    and it is determined by counting the positions where the two strings have different
+    characters.
+
+    Source(s):
+
+        (1) https://en.wikipedia.org/wiki/Hamming_distance
+
+    :param s1, s2: The strings to be analysed
+    :type s1: str
+    :type s2: str
+    :rtype int
+
+    """
+    assert len(s1) == len(s2), "Strings must be the same length."
+    return sum(x[0] != x[1] for x in zip(s1, s2))
+
+
+def lee_distance(s1, s2):
+    """
+    Lee Distance metric calculates the distance between two strings by re-encoding them
+    using a q-ary alphabet of all the unique characters present in one or both strings.
+    Then using the q-ary alphabet index values provides the sum of
+
+        min(| string1_i - string2_i | , alphabet_size - | string1_i - string2_i |)
+
+    for each value i in the length of string1 or string2 (these are interchangeable because
+    both strings must be of equivalent length).
+
+    Source(s):
+
+        (1) https://en.wikipedia.org/wiki/Lee_distance
+
+    :param s1, s2: The strings to be analysed
+    :type s1: str
+    :type s2: str
+    :rtype int
+
+    """
+    assert len(s1) == len(s2), "Strings must be the same length."
+
+    alphabet = set()
+    n1, n2 = len(s1), len(s2)
+    sx = s1 + s2
+    for i in range(n1+n2):
+        alphabet.add(sx[i])
+
+    q = len(alphabet)
+
+    if q >= 2:
+        s1_enc = _lee_string_encoder(s1, alphabet=list(alphabet))
+        s2_enc = _lee_string_encoder(s2, alphabet=list(alphabet))
+        return sum([min([abs(s1_enc[i]-s2_enc[i]), q-abs(s1_enc[i]-s2_enc[i])]) for i in range(len(s1))])
+    else:
+        raise ValueError("number of distinct strings must be greater than 1")
+
+
+def _lee_string_encoder(string, alphabet):
+    for i in range(len(alphabet)):
+        for j in range(len(string)):
+            if alphabet[i] == string[j]:
+                string[j] = i
+    return string
```

Comment:
Is this the expected output of the encoder:

```
>>> alphabet = ['1','2','3','4', '5', '6']
>>> s = list('12112345')
>>> _lee_string_encoder(s, alphabet)
[0, 1, 0, 0, 1, 2, 3, 4]
```

Reply:
Yes. Because it uses the index position of the character value as the encoded value.

Reply:
Could you add some description to the docstring to explain what the encoder is doing?

Reply:
@dtemkin would you mind taking care of this request, then I think we can merge this